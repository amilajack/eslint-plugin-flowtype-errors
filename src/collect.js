/**
 * Run Flow and collect errors in JSON format
 *
 * Reference the following links for possible bug fixes and optimizations
 * https://github.com/facebook/nuclide/blob/master/pkg/nuclide-flow-rpc/lib/FlowRoot.js
 * https://github.com/ptmt/tryflow/blob/gh-pages/js/worker.js
 */
import pathModule from 'path';
import childProcess from 'child_process';
import slash from 'slash';
import filter from './filter';


let flowBin;

try {
  if (!process.env.FLOW_BIN) {
    flowBin = require('flow-bin'); // eslint-disable-line global-require
  }
} catch (e) {
  /* eslint-disable */
  console.log();
  console.log('Oops! Something went wrong! :(');
  console.log();
  console.log('eslint-plugin-flowtype-errors could not find the package "flow-bin". This can happen for a couple different reasons.');
  console.log();
  console.log('1. If ESLint is installed globally, then make sure "flow-bin" is also installed globally.');
  console.log();
  console.log('2. If ESLint is installed locally, then it\'s likely that "flow-bin" is not installed correctly. Try reinstalling by running the following:');
  console.log();
  console.log('  npm i -D flow-bin@latest');
  console.log();
  process.exit(1);
  /* eslint-enable */
}

// Adapted from https://github.com/facebook/flow/blob/master/tsrc/flowResult.js

type FlowLoc = {
  source: ?string
}

type FlowMessage = {
  descr: string,
  type: "Blame" | "Comment",
  loc?: ?FlowLoc
}

type FlowError = {
  message: Array<FlowMessage>,
  operation?: FlowMessage
}

function mainLocOfError(error: FlowError): ?FlowLoc {
  const { operation, message } = error;
  return operation && operation.loc || message[0].loc;
}

/**
 * Wrap critical Flow exception into default Error json format
 */
function fatalError(stderr) {
  return {
    errors: [{
      message: [{
        path: '',
        code: 0,
        line: 0,
        start: 0,
        descr: stderr
      }]
    }]
  };
}

function _formatMessage(message, messages, root, path) {
  switch (message.type) {
    case 'Comment':
      return `${message.descr}`;
    case 'Blame': {
      const see = message.path !== ''
                    ? ` See ${
                      path === message.path
                        ? `line ${message.line}`
                        : `.${slash(message.path.replace(root, ''))}:${message.line}`
                      }.`
                    : '';
      return `'${message.descr}'.${see}`;
    }
    default:
      return `'${message.descr}'.`;
  }
}

function getFlowBin() {
  return process.env.FLOW_BIN || flowBin;
}

function executeFlow(stdin, root, filepath) {
  let stdout;

  switch (stdin && root && filepath && stdin !== '') {
    case true:
      stdout = childProcess.spawnSync(getFlowBin(), [
        'check-contents',
        '--json',
        `--root=${root}`,
        filepath
      ], {
        input: stdin,
        encoding: 'utf-8'
      }).stdout;
      break;
    default:
      stdout = childProcess.spawnSync(getFlowBin(), ['--json']).stdout;
  }

  //
  // This serves as a temporary HACK to prevent 32 bit OS's from failing. Flow does not
  // support 32 bit OS's at the moment.
  // This pretends as if there are now flow errors
  //
  // Ideally, there would be a preinstall npm event to check if the user is on a 32 bit OS
  //

  if (!stdout) {
    return true;
  }

  const stringifiedStdout = stdout.toString();
  let parsedJSONArray;

  try {
    parsedJSONArray = JSON.parse(stringifiedStdout);
  } catch (e) {
    parsedJSONArray = fatalError(stringifiedStdout);
  }

  const fullFilepath = pathModule.resolve(root, filepath);

  // Loop through errors in the file
  const output = parsedJSONArray.errors
    // Temporarily hide the 'inconsistent use of library definitions' issue
    .filter(error => {
      const mainLoc = mainLocOfError(error);
      const mainFile = mainLoc && mainLoc.source;
      return (
        mainFile &&
        error.message[0].descr &&
        !error.message[0].descr.includes('inconsistent use of') &&
        pathModule.resolve(root, mainFile) === fullFilepath
      );
    })
    .map(error => {
      const { message, operation } = error;
      const [firstMessage, ...remainingMessages] = [].concat(operation || [], message);
      const entireMessage = `${firstMessage.descr}: ${
        remainingMessages.reduce(
          (previousMessage,
            currentMessage,
            index,
            messages) =>
            `${previousMessage} ${_formatMessage(
              currentMessage,
              messages,
              root,
              firstMessage.path
            )}`,
          ''
        )
      }`;

      return {
        ...(process.env.DEBUG_FLOWTYPE_ERRRORS === 'true' ? parsedJSONArray : {}),
        message: entireMessage.replace(/\.$/, ''),
        path: firstMessage.path,
        start: firstMessage.loc.start.line,
        end: firstMessage.loc.end.line,
        loc: firstMessage.loc
      };
    });

  return output.length
    ? filter(output)
    : true;
}

export default function collect(stdin, root, filepath) {
  return executeFlow(stdin, root, filepath);
}
