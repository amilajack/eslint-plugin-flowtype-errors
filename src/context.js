/* @flow */

export type EslintContext = {
  getAllComments: Function,
  getFilename: Function,
  getSourceCode: Function,
  report: Function,
  settings: ?{
    'flowtype-errors': ?{
      flowDir?: string,
      stopOnExit?: any
    }
  },
  options: any[]
};
