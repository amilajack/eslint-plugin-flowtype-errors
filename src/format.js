/**
 * This module is responsible for formatting flowtype errors to be more friendly and understandable
 * Formatting should be disabled by default
 * @TODO
 */

function fomatMessage(description) {
  if (description.toLowerCase().includes("' This type")) {
    return description.replace('This type', 'type');
  }

  return description;
}

export default function filter(messages) {
  return messages.map(e => ({
    ...e,
    message: fomatMessage(e.message)
  }));
}
