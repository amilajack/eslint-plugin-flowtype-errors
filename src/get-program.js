// @flow

type Pos = {
  line: number,
  column: number
};

type Loc = {
  start: Pos,
  end: Pos
};

type Program = { text: string, loc: Loc, offset: Pos };

export default function( source: Object, node: Object ): ?Program {

  // Ignore if body is empty.
  if ( node.body.length === 0 ) {
    return;
  }

  const body0 = node.body[0];
  const comments0 = node.comments[0];
  let start;
  let range;

  // With babel-eslint, program.loc.start.column is always 0,
  // workaround it by using the first node of the body to get the offset.

  if ( comments0 ) {
    start = node.range[0] < comments0.range[0] ? {
      line: body0.loc.start.line,
      column: body0.loc.start.column
    } : {
      line: comments0.loc.start.line,
      column: comments0.loc.start.column
    };
    range = [
      Math.min( node.range[0], comments0.range[0] ),
      Math.max( node.range[1], node.comments[node.comments.length - 1].range[1] )
    ];
  } else {
    start = {
      line: body0.loc.start.line,
      column: body0.loc.start.column
    };
    range = node.range;
  }

  return {
    text: source.text.slice( range[0], range[1] ),
    loc: {
      start,
      end: start
    },
    offset: {
      line: start.line - 1,
      column: start.column
    }
  };
}
