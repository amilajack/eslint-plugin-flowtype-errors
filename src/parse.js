/* @flow */

import type { EslintContext } from './context';

export class Script {
  start: number;
  code: string;
  constructor(start: number, code: string = '') {
    this.start = start;
    this.code = code;
  }

  appendLine (line: string) {
    return new Script(this.start, [this.code, line].join(''));
  }
}

function parseCodeBlockLine(line, { scripts, script }) {
  if (/<\/script>/.test(line)) {
    scripts.push(script);
    return { scripts };
  }

  return { scripts, script: script.appendLine(line) };
}

function parseHtmlBlockLine(line, idx, { scripts }) {
  if (/<script>/.test(line)) {
    return { scripts, script: new Script(idx + 1) };
  }

  return { scripts };
}

function parseLine(state, line, idx) {
  if (state.script) {
    return parseCodeBlockLine(line, state);
  }

  return parseHtmlBlockLine(line, idx, state);
}

function extractHtmlScripts(code) {
  const lines = code.match(/[^\r\n]*[\r\n]+/g);

  const res = lines.reduce(parseLine, { scripts: [] });
  return res.scripts;
}

function isHtmlLike(context): boolean {
  return /(\.html)|(\.vue)$/.test(context.getFilename());
}

export function getScripts(context: EslintContext): [Script] {
  const code = context.getSourceCode().getText();

  if (isHtmlLike(context)) {
    return extractHtmlScripts(code);
  }

  return [ new Script(0, code) ];
}
