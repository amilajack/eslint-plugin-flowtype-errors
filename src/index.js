/* @flow */

import path from 'path';
import fs from 'fs';
import { collect, coverage } from './collect';

type EslintContext = {
  getAllComments: Function,
  getFilename: Function,
  getSourceCode: Function,
  report: Function,
  settings: ?{
    "flowtype-errors": ?{
      flowDir?: string,
      stopOnExit?: any
    }
  },
  options: any[]
}

let runOnAllFiles;

function hasFlowPragma(source) {
  return source
    .getAllComments()
    .some(comment => /@flow/.test(comment.value));
}

function lookupFlowDir(context: EslintContext): string {
  const root = process.cwd();
  const flowDirSetting: string = context.settings
    && context.settings['flowtype-errors']
    && context.settings['flowtype-errors'].flowDir || '.';

  return fs.existsSync(path.join(root, flowDirSetting, '.flowconfig'))
    ? path.join(root, flowDirSetting)
    : root;
}

function stopOnExit(context: EslintContext): boolean {
  return !!(context.settings && context.settings['flowtype-errors'] && context.settings['flowtype-errors'].stopOnExit);
}

export default {
  rules: {
    'enforce-min-coverage': function enforceMinCoverage(context: EslintContext) {
      return {
        Program() {
          const source = context.getSourceCode();

          if (hasFlowPragma(source)) {
            const res = coverage(
              source.getText(), lookupFlowDir(context), stopOnExit(context), context.getFilename()
            );

            if (res === true) {
              return;
            }

            const requiredCoverage = context.options[0];
            const { coveredCount, uncoveredCount } = res;

            /* eslint prefer-template: 0 */
            const percentage = Number(Math.round((coveredCount / (coveredCount + uncoveredCount)) * 10000) + 'e-2');

            if (percentage < requiredCoverage) {
              context.report({
                loc: 1,
                message: `Expected coverage to be at least ${requiredCoverage}%, but is: ${percentage}%`
              });
            }
          }
        }
      };
    },
    'show-errors': function showErrors(context: EslintContext) {
      return {
        Program() {
          const source = context.getSourceCode();
          const flowDir = lookupFlowDir(context);

          // Check to see if we should run on every file
          if (runOnAllFiles === undefined) {
            try {
              runOnAllFiles = fs
                .readFileSync(path.join(flowDir, '.flowconfig'))
                .toString()
                .includes('all=true');
            } catch (err) {
              runOnAllFiles = false;
            }
          }

          if (runOnAllFiles === false && !hasFlowPragma(source)) {
            return true;
          }

          const collected = collect(
            source.getText(), flowDir, stopOnExit(context), context.getFilename()
          );

          if (collected === true) {
            return;
          }

          collected.forEach(({ loc, message }) => {
            context.report({
              loc,
              message
            });
          });
        }
      };
    }
  }
};
