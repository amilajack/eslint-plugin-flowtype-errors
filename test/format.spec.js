import path from 'path';
import { expect as chaiExpect } from 'chai';
import { readFileSync, writeFileSync, unlinkSync, mkdirSync } from 'fs';
import { sync as spawnSync } from 'cross-spawn';
import { collect } from '../src/collect';


const testFilenames = [
  '1.example.js',
  '2.example.js',
  '3.example.js',
  '4.example.js',
  '5.example.js',
  '6.example.js',
  '7.example.js',
  '8.example.js',
  '9.example.js'
];

const testResults = testFilenames.map((filename, index) => {
  const root = path.resolve(process.cwd(), 'test');
  const filepath = path.join(root, filename);
  const stdin = readFileSync(filepath).toString();
  const parsedJSONArray = collect(stdin, root, true, filepath);

  return { parsedJSONArray, filename, index };
});

describe('Format', () => {
  for (const { parsedJSONArray, filename } of testResults) {
    it(`${filename} - should have expected properties`, () => {
      const exactFormat = require(`./${filename}`.replace('example', 'expect'));

      chaiExpect(parsedJSONArray).to.be.an('array');

      // Filter out the 'path' property because this changes between environments
      expect(parsedJSONArray.map(e => ({
        end: e.end,
        type: e.type,
        loc: { start: e.loc.start, end: e.loc.end },
        message: e.message,
        start: e.start
      })))
      .toEqual(exactFormat);

      for (const e of parsedJSONArray) {
        if (e !== false) {
          chaiExpect(e.type).to.be.a('string');
          chaiExpect(e.path).to.be.a('string').that.contains(path.join(process.cwd(), 'test'));
          chaiExpect(e.path).to.be.a('string').that.contains('.example.js');
        }
      }
    });
  }
});

const ESLINT_PATH = path.resolve('./node_modules/eslint/bin/eslint.js');

function runEslint(cwd) {
  const result = spawnSync(ESLINT_PATH, ['**/*.js'], { cwd });
  result.stdout = result.stdout && result.stdout.toString();
  result.stderr = result.stderr && result.stderr.toString();
  return result;
}

const codebases = [
  'flow-pragma-1',
  'flow-pragma-2',
  'no-flow-pragma',
  'project-1',
  'run-all',
  'run-all-flowdir',
  'coverage-ok',
  'coverage-ok2',
  'coverage-fail',
  'coverage-fail2'
];

const issue81 = process.platform === 'win32' ? 'folder with spaces' : 'folder-with-spaces';
codebases.push(['folder with spaces', issue81]);
try {
  mkdirSync(path.resolve(`./test/codebases/${issue81}`));
} catch (e) {
  // Already exists
}

const eslintConfig = (enforceMinCoverage) => `
  var Module = require('module');
  var path = require('path');
  var original = Module._resolveFilename;

  // Hack to allow eslint to find the plugin
  Module._resolveFilename = function(request, parent, isMain) {
    if (request === 'eslint-plugin-flowtype-errors') {
      return path.resolve('../../../dist/index.js');
    }
    return original.call(this, request, parent, isMain);
  };

  module.exports = {
    parser: 'babel-eslint',
    root: true, // Make ESLint ignore configuration files in parent folders
    env: {
      node: true,
      es6: true
    },
    plugins: ['flowtype-errors'],
    settings: {
      'flowtype-errors': {
        flowDir: './subdir',
        stopOnExit: 'true'
      }
    },
    rules: {
      ${enforceMinCoverage ? `
        'flowtype-errors/show-errors': 2,
        'flowtype-errors/enforce-min-coverage': [2, ${enforceMinCoverage}]
      ` : `
        'flowtype-errors/show-errors': 2
      `}
    }
  };
`;

describe('Check codebases', () => {
  for (const codebase of codebases) {
    let folder;
    let title;

    if (Array.isArray(codebase)) {
      title = codebase[0];
      folder = codebase[1];
    } else {
      title = codebase;
      folder = codebase;
    }

    it(`${title} - eslint should give expected output`, () => {
      const fullFolder = path.resolve(`./test/codebases/${folder}`);
      const configPath = path.resolve(fullFolder, '.eslintrc.js');

      // Write config file
      writeFileSync(configPath, eslintConfig(folder.match(/^coverage-/) ? 50 : 0));

      // Spawn a eslint process
      const { stdout, stderr } = runEslint(fullFolder);

      const regexp = new RegExp(`^${fullFolder.replace(/[\\^$.*+?()[\]{}|]/g, '\\$&')}.+\\.js$`, 'gm'); // Escape regexp

      // Strip root from filenames
      expect(stdout.replace(regexp, match => match.replace(fullFolder, '.').replace(/\\/g, '/'))).toMatchSnapshot();

      expect(stderr).toEqual('');

      // Clean up
      unlinkSync(configPath);
    });
  }
});
