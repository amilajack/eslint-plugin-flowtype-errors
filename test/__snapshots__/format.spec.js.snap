exports[`Check codebases coverage-fail - eslint should give expected output 1`] = `
"
./example.js
  0:0  error  Expected coverage to be at least 50%, but is: 0%  flowtype-errors/enforce-min-coverage

✖ 1 problem (1 error, 0 warnings)

"
`;

exports[`Check codebases coverage-fail2 - eslint should give expected output 1`] = `
"
./example.js
  0:0  error  Expected coverage to be at least 50%, but is: 33.33%  flowtype-errors/enforce-min-coverage

✖ 1 problem (1 error, 0 warnings)

"
`;

exports[`Check codebases coverage-ok - eslint should give expected output 1`] = `""`;

exports[`Check codebases coverage-ok2 - eslint should give expected output 1`] = `""`;

exports[`Check codebases flow-pragma-1 - eslint should give expected output 1`] = `
"
./example.js
   4:11  error  string:  This type is incompatible with the expected return type of \'boolean\'. See line 3  flowtype-errors/show-errors
   7:7   error  identifier \`x\`:  Could not resolve name                                                    flowtype-errors/show-errors
  14:15  error  string:  The operand of an arithmetic operation must be a number                           flowtype-errors/show-errors

✖ 3 problems (3 errors, 0 warnings)

"
`;

exports[`Check codebases flow-pragma-2 - eslint should give expected output 1`] = `
"
./example.js
   6:11  error  string:  This type is incompatible with the expected return type of \'boolean\'. See line 5  flowtype-errors/show-errors
   9:7   error  identifier \`x\`:  Could not resolve name                                                    flowtype-errors/show-errors
  16:15  error  string:  The operand of an arithmetic operation must be a number                           flowtype-errors/show-errors

✖ 3 problems (3 errors, 0 warnings)

"
`;

exports[`Check codebases folder with spaces - eslint should give expected output 1`] = `""`;

exports[`Check codebases no-flow-pragma - eslint should give expected output 1`] = `""`;

exports[`Check codebases project-1 - eslint should give expected output 1`] = `
"
./1.example.js
   5:11  error  string:  This type is incompatible with the expected return type of \'boolean\'. See line 4  flowtype-errors/show-errors
   8:7   error  identifier \`x\`:  Could not resolve name                                                    flowtype-errors/show-errors
  15:15  error  string:  The operand of an arithmetic operation must be a number                           flowtype-errors/show-errors

./2.example.js
   8:25  error  parameter \`x\`:  Missing annotation       flowtype-errors/show-errors
  12:7   error  identifier \`x\`:  Could not resolve name  flowtype-errors/show-errors

./3.example.js
   9:11  error  string:  This type is incompatible with the expected return type of \'number\'. See line 8   flowtype-errors/show-errors
  13:11  error  string:  This type is incompatible with the expected return type of \'number\'. See line 12  flowtype-errors/show-errors
  16:20  error  string:  This type is incompatible with \'number\'. See line 16                              flowtype-errors/show-errors
  18:8   error  who:  name is already bound \'function who\'. See line 12                                    flowtype-errors/show-errors
  20:12  error  function:  This type cannot be added to \'number\'. See line 20                              flowtype-errors/show-errors

./4.example.js
   8:11  error  property \`lastName\`:  Property not found in \'props of React element \`Foo\`\'. See line 24  flowtype-errors/show-errors
  24:13  error  props of React element \`Foo\`:  This type is incompatible with \'object type\'. See line 8  flowtype-errors/show-errors

./5.example.js
  8:4  error  string:  This type is incompatible with the expected param type of \'number\'. See line 4  flowtype-errors/show-errors

./6.example.js
   8:6  error  number:  This type is incompatible with the expected param type of \'object type\'. See line 4       flowtype-errors/show-errors
  10:2  error  function call:  \'property \`baz\`\'. See line 4. Property not found in \'object literal\'. See line 10  flowtype-errors/show-errors

./7.example.js
  6:4  error  string:  This type is incompatible with the expected param type of \'number\'. See ./5.example.js:4  flowtype-errors/show-errors

./8.example.js
  6:6  error  number:  This type is incompatible with the expected param type of \'object type\'. See ./6.example.js:4      flowtype-errors/show-errors
  8:2  error  function call:  \'property \`baz\`\'. See ./6.example.js:4. Property not found in \'object literal\'. See line 8  flowtype-errors/show-errors

./9.example.js
  8:18  error  React element \`Hello\`:  \'property \`name\`\'. See ./9.example.import.js:6. Property not found in \'props of React element \`Hello\`\'. See line 8  flowtype-errors/show-errors

✖ 19 problems (19 errors, 0 warnings)

"
`;

exports[`Check codebases run-all - eslint should give expected output 1`] = `
"
./example.js
   2:11  error  string:  This type is incompatible with the expected return type of \'boolean\'. See line 1  flowtype-errors/show-errors
   5:7   error  identifier \`x\`:  Could not resolve name                                                    flowtype-errors/show-errors
  12:15  error  string:  The operand of an arithmetic operation must be a number                           flowtype-errors/show-errors

✖ 3 problems (3 errors, 0 warnings)

"
`;

exports[`Check codebases run-all-flowdir - eslint should give expected output 1`] = `
"
./subdir/example.js
   2:11  error  string:  This type is incompatible with the expected return type of \'boolean\'. See line 1  flowtype-errors/show-errors
   5:7   error  identifier \`x\`:  Could not resolve name                                                    flowtype-errors/show-errors
  12:15  error  string:  The operand of an arithmetic operation must be a number                           flowtype-errors/show-errors

✖ 3 problems (3 errors, 0 warnings)

"
`;
